java/pom.xml

```
@@ -905,6 +905,69 @@
       </build>
     </profile>
 
+    <profile>
+      <id>generate-jnicpp-dylib_so</id>
+      <properties>
+        <ARROW_DATASET>OFF</ARROW_DATASET>
+        <ARROW_GANDIVA>OFF</ARROW_GANDIVA>
+        <ARROW_ORC>OFF</ARROW_ORC>
+        <arrow.cpp.build.dir>../java-dist</arrow.cpp.build.dir>
+      </properties>
+      <build>
+        <plugins>
+          <plugin>
+            <groupId>org.codehaus.mojo</groupId>
+            <artifactId>exec-maven-plugin</artifactId>
+            <version>3.1.0</version>
+            <executions>
+              <execution>
+                <id>jnicppdefine</id>
+                <phase>generate-resources</phase>
+                <goals>
+                  <goal>exec</goal>
+                </goals>
+                <configuration>
+                  <executable>cmake</executable>
+                  <commandlineArgs>
+                    -DARROW_*_USE_SHARED=OFF
+                    -DARROW_JNI=ON
+                    -DARROW_PARQUET=ON
+                    -DARROW_FILESYSTEM=ON
+                    -DARROW_DATASET=${ARROW_DATASET}
+                    -DARROW_GANDIVA_JAVA=ON
+                    -DARROW_GANDIVA_STATIC_LIBSTDCPP=ON
+                    -DARROW_GANDIVA=${ARROW_GANDIVA}
+                    -DARROW_ORC=${ARROW_ORC}
+                    -DCMAKE_BUILD_TYPE=Release
+                    -DCMAKE_INSTALL_LIBDIR=lib
+                    -DCMAKE_INSTALL_PREFIX=${arrow.cpp.build.dir}
+                    -DCMAKE_UNITY_BUILD=ON
+                    -DARROW_DEPENDENCY_SOURCE=BUNDLED
+                    ../cpp
+                  </commandlineArgs>
+                  <workingDirectory>../java-native-cpp</workingDirectory>
+                </configuration>
+              </execution>
+              <execution>
+                <id>jnicppbuild</id>
+                <phase>generate-resources</phase>
+                <goals>
+                  <goal>exec</goal>
+                </goals>
+                <configuration>
+                  <executable>cmake</executable>
+                  <commandlineArgs>
+                    --build . --target install --config release
```

This will work but `Release` is better because we specify `-DCMAKE_BUILD_TYPE=Release`:

```suggestion
                    --build . --target install --config Release
```